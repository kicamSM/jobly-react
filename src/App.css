body { background-color: #e5e5f7;
opacity: 0.8;
background-image: linear-gradient(-45deg, #e5e5f7, #e5e5f7 50%, #444cf7 50%, #444cf7);
background-size: 10px 10px;
}

.App {
  text-align: center;
}


.App-header {
  background-color: #282c34;
  min-height: 100vh;
  display: flex;
  flex-direction: column;
  align-items: center;
  justify-content: center;
  font-size: calc(10px + 2vmin);
  color: white;
}

.App-link {
  color: #61dafb;
}


