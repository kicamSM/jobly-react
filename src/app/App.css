
.App {
  text-align: center;
  justify-content: center;

}

body {
  background-image: url("./image/background.jpg");
  background-size: cover;
}


