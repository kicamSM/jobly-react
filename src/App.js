import { BrowserRouter, Redirect } from "react-router-dom";
import './App.css';
import NavBar from "./navbar/NavBar";
import Routes from './Routes';
import JoblyApi from "./Api";
import React, { useState, useEffect } from 'react';
import useLocalStorage from './hooks/useLocalStorage';
import { useHistory } from "react-router-dom/cjs/react-router-dom.min";
import jwt from "jsonwebtoken";
import UserContext from "./UserContext";

export const TOKEN_STORAGE_ID = "jobly-token";

function App() {

  const [user, setUser] = useState(null);
  const [token, setToken] = useState([]);
  // const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);
  // ! come back to this useLocalStorage
  const [applicationIds, setApplicationIds] = useState(new Set([]));
  const history = useHistory();
  console.log("history!!!!:", history)

// useEffect(() => {
//   const setTokenToLocalStorage = (token) => {
//     const stringToken = JSON.stringify(token);
//     useLocalStorage.setItem('token', stringToken);
//   };
//   }, [token]);

//   useEffect(() => {
//     const setUserToLocalStorage = (token) => {
//       const stringToken = JSON.stringify(token);
//       useLocalStorage.setItem('token', stringToken);
//     };
//     }, [token]);


  // useEffect(() => {
  //   async function getUser() {
  //     let user = await JoblyApi.getUser();
  //     console.log('user in Routes', user)
  //     setUser(user);
  //     // setIsLoading(false);
  //   }
  //   getUser();
  // }, []);

  // console.log("user:", user)

  // useEffect(() => {
  //   async function addUser(userData) {
  //     let user = await JoblyApi.addUser(userData);
  //     console.log('user in Routes', user)
  //     addUser();
  //     // setIsLoading(false);
  //   }
  //   addUser();
  // }, []);

 // Load user info from API. Until a user is logged in and they have a token,
  // this should not run. It only needs to re-run when a user logs out, so
  // the value of the token is a dependency for this effect.

  useEffect(function loadUser() {
    console.debug("App useEffect loadUserInfo", "token=", token);

    async function getUser() {
      if (token) {
        try {
          let { username } = jwt.decode(token);
          // put the token on the Api class so it can use it to call the API.
          console.log("username:", username)
          JoblyApi.token = token;
          let user = await JoblyApi.getUser(username);
          console.log("user:", user)
          console.log("user.firstName:", user.firstName)
          setUser(user);
          setApplicationIds(new Set(user.applications));
        } catch (err) {
          console.error("App loadUserInfo: problem loading", err);
          setUser(null);
        }
      }
      // setInfoLoaded(true);
      console.log("setinfoloaded true")
    }

    // set infoLoaded to false while async getCurrentUser runs; once the
    // data is fetched (or even if an error happens!), this will be set back
    // to false to control the spinner.
    // setInfoLoaded(false);
    console.log("setinfoloaded false")
    getUser();
  }, [token]);

  async function signup(userData) {
    let token = JoblyApi.signup(userData); 
    setToken(token);
  }

  async function login(loginData) {
    try {
      let token = await JoblyApi.login(loginData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("login failed", errors);
      return { success: false, errors };
    }
  }

  function logout() {
    // try {
      console.log("token in logout before setting token:", token)
      setToken(null);
      console.log("token in logout after setting token:", token)
      // history.push('/login');
      // redirect()
      // return <Redirect to="/login" />;
      // return { success: true };
      // return <Navigate to="/login" replace={true} />;
    // } catch (errors) {
    //   console.error("logout failed", errors);
    //   return { success: false, errors };
    // }

  }

  async function update(username, data) {
    try {
      let token = await JoblyApi.update(username, data);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("update failed", errors);
      return { success: false, errors };
    }
  }

console.log("user!!!!:", user)

  return (
    <div className="App">
      <BrowserRouter>
        <UserContext.Provider value={{user, setUser}}>
        {/* <NavBar logout={logout} history={history}/> */}
        <NavBar logout={logout} user={user} />
        <main>
          <Routes login={login} signup={signup} update={update} />
        </main>
        </UserContext.Provider>
      </BrowserRouter>
    </div>
  );
}

export default App;
