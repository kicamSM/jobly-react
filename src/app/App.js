import React, { useState, useEffect } from 'react';
import { BrowserRouter } from "react-router-dom";
import NavBar from "../navbar/NavBar";
import Routes from "../routes/Routes";
import UserContext from "../repeated/UserContext";
import JoblyApi from "../Api";
import Loading from "../repeated/loading/Loading"
import './App.css';
import useLocalStorage from "../hooks/useLocalStorage";
import jwt from "jsonwebtoken";

export const TOKEN_STORAGE_ID = "jobly-token";

  /**  
  * App
  */

function App() {

   /** Set user and application ids in state, set token in local storage, and set history*/


  const [user, setUser] = useState(null);
  const [token, setToken] = useLocalStorage(TOKEN_STORAGE_ID);
  const [isLoading, setIsLoading] = useState(true);


  useEffect(function loadUser() {
    console.debug("App useEffect loadUserInfo", "token=", token);

    /** If token then API get request for user using username after decoding token if value of token changes rerun. */

    async function getUser() {
      if (token) {

        try {
          let { username } = jwt.decode(token);
          JoblyApi.token = token;
          let user = await JoblyApi.getUser(username);
          setUser(user);
        } catch (err) {
          console.error("App loadUserInfo: problem loading", err);
          setUser(null);
        }
      }
      setIsLoading(false);
      console.log("setinfoloaded true")
    }
    getUser();
  }, [token]);

  /** API post request for user signup or return error */

  async function signup(userData) {
    try {
    let token = await JoblyApi.signup(userData); 
    setToken(token);
    return { success: true};
    } catch (errors) {
      console.log("signup failed", errors);
      return {success: false, errors};
    }
  }

  /** API post request for user login or return error */

  async function login(loginData) {
    try {
      let token = await JoblyApi.login(loginData);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("login failed", errors);
      return { success: false, errors };
    }
  }

  /** Logout - clear token and user */

  function logout() {
      setToken(null);
      setUser(null);
  }

  /** API patch request for user update or return error */

  async function update(username, data) {
    try {
      let token = await JoblyApi.update(username, data);
      setToken(token);
      return { success: true };
    } catch (errors) {
      console.error("update failed", errors);
      return { success: false, errors };
    }
  }

  /** API post request for job application or return error */

async function apply(username, id) {
  try {
    let token = await JoblyApi.apply(username, id);
    setToken(token);
    return { success: true };
  } catch (errors) {
    console.error("applied to job failed", errors);
    return { success: false, errors };
  }
}

  /** Display isLoading if API call is has not returned */

  if (isLoading) {
    return (
        <Loading />
    )
  }

  /** Render App */

  return (
    <div className="App">
      <BrowserRouter>
        <UserContext.Provider value={{user, setUser}}>
          <NavBar logout={logout} />
          <main>
            <Routes login={login} signup={signup} update={update} apply={apply} />
          </main>
        </UserContext.Provider>
      </BrowserRouter>
    </div>
  );
}

export default App;
