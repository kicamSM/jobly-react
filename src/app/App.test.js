import React from 'react';
import { render, screen } from '@testing-library/react';
import App from './App';
import { toMatchSnapshot } from 'jest-snapshot'

expect.extend({ toMatchSnapshot });

test('App renders without crashing', () => {
  render(<App />);
});

// test('matches snapshot', () => {
//   const {asFragment} = render(<App />);
//   expect(asFragment()).toMatchSnapShot(); 
// })

it('matches snapshot', () => {
  const {asFragment} = render(<App />);
  expect(asFragment()).toMatchSnapShot(); 
});

