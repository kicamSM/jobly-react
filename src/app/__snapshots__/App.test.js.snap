// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="App"
  >
    <div>
      <nav
        class="navbar navbar-expand-md"
      >
        <div
          class="container-fluid"
        >
          <a
            aria-current="page"
            class="navbar-brand active"
            href="/"
          >
            Jobly
          </a>
          <ul
            class="ml-auto navbar-nav"
          >
            <li
              class="Navbar-Signup nav-item"
            >
              <a
                href="/signup"
              >
                Signup
              </a>
            </li>
            <li
              class="Navbar-Login nav-item"
            >
              <a
                href="/login"
              >
                Login
              </a>
            </li>
          </ul>
        </div>
      </nav>
    </div>
    <main>
      <section
        class="col-md-8 Home"
      >
        <div
          class="card"
        >
          <div
            class="card-title"
          >
            <h1>
              Jobly
            </h1>
          </div>
          <div
            class="card-body"
          >
            <p
              class="card-text"
            >
              All the jobs in one, convenient place.
            </p>
             
            <p
              class="card-text"
            >
              <a
                class="Home-Link"
                href="/login"
              >
                <button
                  class="btn btn-secondary"
                >
                  Login
                </button>
              </a>
              <a
                class="Home-Link"
                href="/signup"
              >
                <button
                  class="btn btn-secondary"
                >
                  Signup
                </button>
              </a>
            </p>
          </div>
        </div>
      </section>
    </main>
  </div>
</DocumentFragment>
`;
